File: core/src/main/scala/diplomacy/variant/standard/rule/BuildValidator.scala
package diplomacy.variant.standard.rule

import scala.collection.{ mutable => cmutable }

import diplomacy.rule.InvalidOrderMessage
import diplomacy.board.Power

class BuildValidator[Turn_ <: Turn, Power_ <: Power] extends Validator[Turn_, Power_] {
  def unitsRequiringOrder(board: Board): Set[DiplomacyUnit] = Set()

  def errorMessageOfOrder(board: Board)(order: Order): Option[InvalidOrderMessage] = {
    order match {
      case Order.Build(unit) =>
        if (board.units exists { _.location.province == unit.location.province }) {
          Option(InvalidOrderMessage(s"An unit is in ${unit.location.province.name}."))
        } else if (unit.location.province.homeOf != Option(unit.power)) {
          Option(InvalidOrderMessage(s"${unit.power} cannot build an unit in ${unit.location}."))
        } else if (!unit.location.province.isSupplyCenter) {
          Option(InvalidOrderMessage(s"${unit.location.province.name} is not supply center."))
        } else if (board.occupation.get(unit.location.province) != Some(unit.power)) {
          Option(
            InvalidOrderMessage(s"${unit.location.province.name} is not occupied by ${unit.power}.")
          )
        } else {
          None
        }
      case Order.Disband(unit) =>
        if (!(board.units contains unit)) {
          Option(InvalidOrderMessage(s"${unit} does not exist."))
        } else {
          val numOfUnits = board.units count { _.power == unit.power }
          val numOfSupplys = board.numberOfSupplyCenters.getOrElse(unit.power, 0)
          if (numOfUnits <= numOfSupplys) {
            Option(InvalidOrderMessage(s"${unit.power} has sufficient supply centers."))
          } else {
            None
          }
        }
      case _ => Option(InvalidOrderMessage(s"${order} is not order for build phase."))
    }
  }
  def errorMessageOfOrders(board: Board)(orders: Set[Order]): Option[InvalidOrderMessage] = {
    board.map.powers find { power =>
      val numOfUnits = board.units count { unit => power == unit.power }
      val numOfSupplys = board.numberOfSupplyCenters.getOrElse(power, 0)

      val diff = (orders map {
        case Order.Build(unit) if unit.power == power => 1
        case Order.Disband(unit) if unit.power == power => -1
        case _ => 0
      }).sum

      (numOfUnits + diff) > numOfSupplys
    } map { power => InvalidOrderMessage(s"${power} does not have enough supply centers.") }
  }
}

File: src/variant/standard/rule/build-validator.ts
import { Validator } from "./validator"
import { StandardRuleUtils } from "./standard-rule-utils"
import { OrderType, Order } from "./order"
import { Board, Unit } from "./types"

export class BuildValidator<Power> implements Validator<Power> {
  /**
   * @param stringify Stringify instances of Power
   */
  constructor (private stringify: (power: Power) => String) {}

  unitsRequiringOrder (board: Board<Power>) {
    return new Set()
  }

  errorMessageOfOrder (board: Board<Power>, order: Order<Power>) {
    const numberOfSupplyCenters = StandardRuleUtils.numberOfSupplyCenters(board)
    switch (order.tpe) {
      case OrderType.Build:
        if ([...board.units].some(unit => unit.location.province === order.unit.location.province)) {
          return `An unit is in ${order.unit.location.province.name}`
        } else if (order.unit.location.province.homeOf !== order.unit.power) {
          return `${this.stringify(order.unit.power)} cannot build an unit in ${order.unit.location}`
        } else if (!order.unit.location.province.isSupplyCenter) {
          return `${order.unit.location.province.name} is not supply center`
        } else {
          const status = board.provinceStatuses.get(order.unit.location.province)
          if (!status || status.occupied !== order.unit.power) {
            return `${order.unit.location.province.name} is not occupied by ${this.stringify(order.unit.power)}`
          }
        }
        break
      case OrderType.Disband:
        if (!board.units.has(order.unit)) {
          return `${order.unit} does not exist`
        }
        const numOfUnits =
          ([...board.units].filter(unit => unit.power === order.unit.power)).length
        const numOfSupply = numberOfSupplyCenters.get(order.unit.power) || 0
        if (numOfUnits <= numOfSupply) {
          return `${this.stringify(order.unit.power)} has sufficient supply centers`
        }
        break
      default:
        return `${order} is not for build phase`
    }
    return null
  }

  errorMessageOfOrders (board: Board<Power>, orders: Set<Order<Power>>) {
    const numberOfSupplyCenters = StandardRuleUtils.numberOfSupplyCenters(board)
    const power = [...board.map.powers].find(power => {
      const numOfUnits = ([...board.units].filter(unit => unit.power === power)).length
      const numOfSupply = numberOfSupplyCenters.get(power) || 0
      const diffs = [...orders].map(order => {
        if (order.tpe === OrderType.Build && order.unit.power === power) {
          return 1
        } else if (order.tpe === OrderType.Disband && order.unit.power === power) {
          return -1
        } else {
          return 0
        }
      })
      const diff = diffs.reduce((prev, curr) => prev + curr, 0)
      return (numOfUnits + diff) > numOfSupply
    })

    if (power) {
      return `${this.stringify(power)} does not have enough supply centers`
    }
    return null
  }
}
